### PR TITLE
[pulsar-function-go] Add statistics and Prometheus to Go Function instances for production readiness

### PR DESCRIPTION
**Pull request by devinbost**
Master issue: #4175
This PR also depends on the code in #6031 .

This PR is to add statistics (to enable calls to get function status, get function stats, and get other metrics) for Pulsar Admin usage and for sending to Prometheus. 
Because Prometheus Go did not support some of the ways that we depended on to interact with Prometheus in the Python and Java Pulsar code, this change was more significant and diverges slightly from the approaches used for the other parts of the Pulsar codebase. However, based on feedback from the core maintainers of Prometheus, there is a risk that they may deprecate the other ways that we're using their library. So, the approaches used in this PR should serve as the new standard.  For more information about the issues with Prometheus, please see the discussion here: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/prometheus-users/NpkERPC17H4

Fixes issue #6205 

### Motivation

The current implementation of Go functions in Pulsar is missing key statistical information and other information that is critical for determining the health and reliability of Go functions in production. This PR solves that problem. 

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API: (sort of)
It adds functionality that didn't previously exist for Go functions.
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (sort of)
It adds functionality that didn't previously exist for Go functions.
  - The admin cli options: (sort of)
It adds functionality that didn't previously exist for Go functions.
  - Anything that affects deployment: (possibly, though unlikely, due to new dependencies)

### Documentation

  - Does this pull request introduce a new feature? (sort of)
Prometheus metrics were changed slightly. We needed to replace many of the Prometheus Counters with Gauges due to limitations in the Prometheus Go client. 
  - If yes, how is the feature documented? (not documented)
  - If a feature is not applicable for documentation, explain why?
More documentation will need to be created as Go functions become more production-ready.

### CODE REVIEW

**Comment by devinbost** [comment]
@sijie This is ready now for code review.

**Comment by jiazhai** [comment]
@sijie @wolfstudy Would you please help review this issue.
@devinbost Thanks for the fix. seems master branch updating brings in some conflicts, would you please help rebase this work.

**Comment by devinbost** [comment]
@jiazhai It will be easier for me to rebase this if we first merge this one: https://github.com/apache/pulsar/pull/6104

**Comment by devinbost** [comment]
@jerrypeng @merlimat Please note that in this PR, I'm using the new approach to get data from Prometheus (based on their recommendations) because the way we were doing it in Python and Java is not supported in their newer client libraries. (They were very adamant about it too.)

**Comment by devinbost** [comment]
The test failures in this PR are just flaky tests. Functionally, everything works, along with https://github.com/apache/pulsar/pull/6104
(It's just not very effective for me to need to push a new commit every time I want to re-run the tests because every time I do that, new failures occur. Hence, why I've been working on addressing the flaky tests in https://github.com/apache/pulsar/pull/6202 )

**Comment by devinbost** [comment]
@sijie @merlimat @jerrypeng This PR will need a more careful review to determine if I'm incrementing the counters and gauges (and other similar Prometheus objects) at the right locations. I was a little uncertain about some of them.

File: pulsar-function-go/pf/stats_test.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	"math"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/prometheus/client_golang/prometheus"
+	io_prometheus_client "github.com/prometheus/client_model/go"
+	"github.com/stretchr/testify/assert"
+)
+
+/*func test(){
```
Comment (jiazhai) [inline pulsar-function-go/pf/stats_test.go:33]: Could this be removed?

File: pulsar-function-go/pf/stats.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	errorTs := LatestException{exception, ts}
+	stat.latestSysException = append(stat.latestSysException, errorTs)
+	if len(stat.latestSysException) > 10 {
+		stat.latestSysException = stat.latestSysException[1:]
+	}
+	// report exception via prometheus
+	stat.reportSystemExceptionPrometheus(exception, ts)
+}
+
+//@limits(calls=5, period=60)
```
Comment (jiazhai) [inline pulsar-function-go/pf/stats.go:292]: Could this be removed?

File: pulsar-function-go/pf/stats.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+	errorTs := LatestException{err, ts}
+	stat.latestUserException = append(stat.latestUserException, errorTs)
+	if len(stat.latestUserException) > 10 {
+		stat.latestUserException = stat.latestUserException[1:]
+	}
+	// report exception via prometheus
+	stat.reportUserExceptionPrometheus(err, ts)
+}
+
+//@limits(calls=5, period=60)
```
Comment (jiazhai) [inline pulsar-function-go/pf/stats.go:262]: could this be removed?

File: pulsar-function-go/pf/instance.go [hunk 4]
```diff
@@ -374,17 +409,201 @@ func (gi *goInstance) healthCheck() *pb.HealthCheckResult {
...
+
+func (gi *goInstance) getTotalUserExceptions1min() float32 {
+	metric := gi.getMatchingMetricFromRegistry(PulsarFunctionMetricsPrefix + TotalUserExceptions1min)
+	// "pulsar_function_" + "user_exceptions_total_1min", GaugeVec
+	val := metric.GetGauge().Value
+	return float32(*val)
+}
+
+/*
+func (gi *goInstance) get_avg_process_latency_1min() float32 {
```
Comment (jiazhai) [inline pulsar-function-go/pf/instance.go:592]: Do we still need this?

**Comment by jiazhai** [comment]
@devinbost thanks for the great work. Left some minor comments. overall. lgtm.
It would be more great if you could provide some documents on this?
@sijie @wolfstudy Would you please take a look of this PR?

File: pulsar-function-go/pf/instance.go (hunk 4 above)
Comment (devinbost) [reply pulsar-function-go/pf/instance.go:592]: @jiazhai This is a method that isn't available due to the Prometheus architecture change they rolled out to their Go library. So, we need to find another way to capture the 1 min metrics. The comment was to note that this feature is currently missing.

**Comment by devinbost** [comment]
> @devinbost thanks for the great work. Left some minor comments. overall. lgtm.
> It would be more great if you could provide some documents on this?
> @sijie @wolfstudy Would you please take a look of this PR?

Thanks for the feedback. What kind of documents would be helpful?

**Comment by devinbost** [comment]
@jiazhai 
> @devinbost thanks for the great work. Left some minor comments. overall. lgtm.
> It would be more great if you could provide some documents on this?
> @sijie @wolfstudy Would you please take a look of this PR?

Thanks for the feedback. What kind of documents would be helpful?

File: pulsar-function-go/pf/instance.go [hunk 5]
```diff
@@ -128,20 +146,27 @@ CLOSE:
...
 				if autoAck && atLeastOnce {
 					gi.nackInputMessage(msgInput)
 				}
+				gi.stats.incrTotalUserExceptions(err)
 				return err
 			}
 
 			gi.processResult(msgInput, output)
 
+			gi.stats.processTimeEnd() // Should this be called here or before processResult(..)?
```
Comment (sijie) [inline pulsar-function-go/pf/instance.go:167]: I think this should be called before `processResult`.

File: pulsar-function-go/pf/stats_test.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+	// Just for demonstration, let's check the state of the summary vector
+	// by registering it with a custom registry and then let it collect the
+	// metrics.
+	reg := prometheus.NewRegistry()
+	reg.MustRegister(temps)
+
+	metricFamilies, err := reg.Gather()
+	if err != nil || len(metricFamilies) != 1 {
+		panic("unexpected behavior of custom test registry")
```
Comment (wolfstudy) [inline pulsar-function-go/pf/stats_test.go:92]: Please replace `panic` with `t.Fatal()`

File: pulsar-function-go/pf/stats_test.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+	}
+	match := func(vect *io_prometheus_client.MetricFamily) bool {
+		return *vect.Name == "pond_temperature_celsius"
+	}
+	fiteredMetricFamilies := filter(metricFamilies, match)
+
+	if len(fiteredMetricFamilies) > 1 {
+		panic("Too many metric families")
```
Comment (wolfstudy) [inline pulsar-function-go/pf/stats_test.go:100]: Please replace `panic` with `t.Fatal()`

File: pulsar-function-go/pf/stats_test.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+	}
+	// Then, we need to filter the metrics in the family to one that matches our label.
+
+	fmt.Println(proto.MarshalTextString(metricFamilies[0]))
```
Comment (wolfstudy) [inline pulsar-function-go/pf/stats_test.go:104]: Why don't we use assert to compare whether the expected value and the actual value are equal? If my understanding is correct, this test case will succeed at any time, right?

E.g:
```
assert.Equal(t, x, y)
```

File: pulsar-function-go/pf/stats_test.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+	//     quantile: <
+	//       quantile: 0.9
+	//       value: 41.3
+	//     >
+	//     quantile: <
+	//       quantile: 0.99
+	//       value: 41.9
+	//     >
+	//   >
+	// >
```
Comment (wolfstudy) [inline pulsar-function-go/pf/stats_test.go:175]: Please remove these comments if we don't need

**Comment by devinbost** [comment]
I apologize for not getting the comments addressed yet. I've had a lot of other responsibilities lately.

**Comment by Anonymitaet** [comment]
Have discussed w/ @wolfstudy, docs have been added [here](https://streamnative.io/docs/latest/process/pulsar-functions/deploy/monitor). 
Will copy to Pulsar later.

**Pull request merged**